library/include/ck/library/host/host_interface.hpp

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "config.hpp"
+
+#include "device_base.hpp"
+
+struct DeviceConvFwdPtr_t
+{
+	using BaseArgument = ck::tensor_operation::device::BaseArgument;
+	using BaseInvoker = ck::tensor_operation::device::BaseInvoker;
+
+	struct DeviceConvFwdPtrImpl;
+	std::unique_ptr<DeviceConvFwdPtrImpl> pImpl;
+	DeviceConvFwdPtr_t();
+	~DeviceConvFwdPtr_t();
+	DeviceConvFwdPtr_t(DeviceConvFwdPtr_t&&);
+	DeviceConvFwdPtr_t(DeviceConvFwdPtrImpl&);
+	DeviceConvFwdPtr_t& operator=(DeviceConvFwdPtr_t&) = delete;
+	DeviceConvFwdPtr_t& operator=(const DeviceConvFwdPtr_t&)=delete;
+	std::unique_ptr<BaseArgument> MakeArgumentPointer(void* in_ptr, void* wei_ptr, void* out_ptr, 
+				 size_t N, size_t K, size_t C, 
+				 std::vector<ck::index_t> input_spatial_lengths,
+				 std::vector<ck::index_t> filter_spatial_lengths,
+				 std::vector<ck::index_t> output_spatial_lengths,
+				 std::vector<ck::index_t> conv_filter_strides,
+				 std::vector<ck::index_t> conv_filter_dilations,
+				 std::vector<ck::index_t> input_left_pads,
+				 std::vector<ck::index_t> input_right_pads) const; // in,wei and out element ops are ignored for now since even if we change them, they cant be linked
+	std::unique_ptr<BaseInvoker> MakeInvokerPointer() const; // requires including BaseInvoker headers
+	std::string GetTypeString();
+	bool IsSupportedArgument(const BaseArgument* arg_ptr);
+};
```

Comment:
There are a couple of issues with the header you mentioned above. Firstly, it is templated on the input and output types which can create issues for a library that will include it. In the `ckProfiler` and elsewhere the device/host headers are included which means that the compiler can instantiate those objects and compile them. For a library that is not possible since it should not depend on instantiating code. Furthermore, the `OpInstanceRunEngine` class has functions implemented in the header file which means the exact implementation of the referenced objects would need to be exposed to MIOpen. This again can create multiple issues when included in another library.  Thirdly, some methods are pure virtual which also causes issues as in the derived classes are not known and the compiler optimizes things out. This however, may be a minor issue.

However, all is not lost. I recommend the above two classes inherit from the interface I created and then you can freely use metaprogramming to specialize types and still maintain a common interface while at the same time hiding all the implementation which is the primary purpose here. We can discuss this further in our usual meetings.